# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/src/vm/EnvironmentObject.cpp
# Commit: c42e89b08392
# Full Hash: c42e89b083921095db6967a37ffb1256a8e5598d
# Author: Tooru Fujisawa <arai_a@mac.com>
# Date: 2025-07-02 16:04:25
# Regressor Bug: 1970388
# File Overlap Count: 1
# Description:
#   Bug 1970388 - Part 2: Reflect the enclosing environment to MissingEnvironmentMap. r=bthrall
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D253505
# ==============================================================================

diff -r d4c97004a4a1 -r c42e89b08392 js/src/vm/EnvironmentObject.cpp
--- a/js/src/vm/EnvironmentObject.cpp	Wed Jul 02 11:03:28 2025 +0000
+++ b/js/src/vm/EnvironmentObject.cpp	Wed Jul 02 11:03:28 2025 +0000
@@ -1553,15 +1553,39 @@
   return false;
 }
 
+MissingEnvironmentKey::MissingEnvironmentKey(JSContext* cx,
+                                             const EnvironmentIter& ei)
+    : frame_(ei.maybeInitialFrame()),
+      nearestEnv_(nullptr),
+      scope_(ei.maybeScope()) {
+  if (!frame_) {
+    EnvironmentIter copy(cx, ei);
+    while (copy) {
+      if (copy.hasAnyEnvironmentObject()) {
+        nearestEnv_ = &copy.environment();
+        break;
+      }
+      ++copy;
+    }
+
+    // The global object should have an environment object even if we don't find
+    // anything else.
+    MOZ_ASSERT(nearestEnv_.unbarrieredGet());
+  }
+}
+
 /* static */
 HashNumber MissingEnvironmentKey::hash(MissingEnvironmentKey ek) {
-  return size_t(ek.frame_.raw()) ^ size_t(ek.scope_);
+  return size_t(ek.frame_.raw()) ^ size_t(ek.nearestEnv_.unbarrieredGet()) ^
+         size_t(ek.scope_);
 }
 
 /* static */
 bool MissingEnvironmentKey::match(MissingEnvironmentKey ek1,
                                   MissingEnvironmentKey ek2) {
-  return ek1.frame_ == ek2.frame_ && ek1.scope_ == ek2.scope_;
+  return ek1.frame_ == ek2.frame_ &&
+         ek1.nearestEnv_.unbarrieredGet() == ek2.nearestEnv_.unbarrieredGet() &&
+         ek1.scope_ == ek2.scope_;
 }
 
 bool LiveEnvironmentVal::traceWeak(JSTracer* trc) {
@@ -2678,12 +2702,28 @@
       liveEnvs.remove(&result.initialTarget()->environment());
       e.removeFront();
     } else {
+      bool needsRekey = false;
+
       MissingEnvironmentKey key = e.front().key();
       Scope* scope = key.scope();
       MOZ_ALWAYS_TRUE(TraceManuallyBarrieredWeakEdge(
           trc, &scope, "MissingEnvironmentKey scope"));
       if (scope != key.scope()) {
         key.updateScope(scope);
+
+        needsRekey = true;
+      }
+
+      EnvironmentObject* oldEnv = key.nearestEnvUnbarriered();
+      if (oldEnv) {
+        TraceWeakEdge(trc, &key.nearestEnvRaw(),
+                      "MissingEnvironmentKey nearestEnv");
+        if (oldEnv != key.nearestEnvUnbarriered()) {
+          needsRekey = true;
+        }
+      }
+
+      if (needsRekey) {
         e.rekeyFront(key);
       }
     }
@@ -2709,6 +2749,7 @@
    */
   CheckTableAfterMovingGC(missingEnvs, [this](const auto& entry) {
     CheckGCThingAfterMovingGC(entry.key().scope(), zone());
+    CheckGCThingAfterMovingGC(entry.key().nearestEnvUnbarriered(), zone());
     // Use unbarrieredGet() to prevent triggering read barrier while collecting.
     CheckGCThingAfterMovingGC(entry.value().unbarrieredGet(), zone());
     return entry.key();
@@ -2799,7 +2840,7 @@
   }
 
   if (MissingEnvironmentMap::Ptr p =
-          envs->missingEnvs.lookup(MissingEnvironmentKey(ei))) {
+          envs->missingEnvs.lookup(MissingEnvironmentKey(cx, ei))) {
     MOZ_ASSERT(CanUseDebugEnvironmentMaps(cx));
     return p->value();
   }
@@ -2822,7 +2863,7 @@
     return false;
   }
 
-  MissingEnvironmentKey key(ei);
+  MissingEnvironmentKey key(cx, ei);
   MOZ_ASSERT(!envs->missingEnvs.has(key));
   if (!envs->missingEnvs.put(key,
                              WeakHeapPtr<DebugEnvironmentProxy*>(debugEnv))) {
@@ -3031,7 +3072,7 @@
 
   Rooted<Environment*> env(cx);
   if (MissingEnvironmentMap::Ptr p =
-          envs->missingEnvs.lookup(MissingEnvironmentKey(ei))) {
+          envs->missingEnvs.lookup(MissingEnvironmentKey(cx, ei))) {
     env = &p->value()->environment().as<Environment>();
     envs->missingEnvs.remove(p);
   } else if (ei.hasSyntacticEnvironment()) {
@@ -3221,6 +3262,16 @@
   for (MissingEnvironmentMap::Enum e(missingEnvs); !e.empty(); e.popFront()) {
     if (e.front().key().frame() == frame) {
       TraceEdge(trc, &e.front().value(), "debug-env-live-frame-missing-env");
+
+      MissingEnvironmentKey key = e.front().key();
+      EnvironmentObject* oldEnv = key.nearestEnvUnbarriered();
+      if (oldEnv) {
+        TraceWeakEdge(trc, &key.nearestEnvRaw(),
+                      "MissingEnvironmentKey nearestEnv");
+        if (oldEnv != key.nearestEnvUnbarriered()) {
+          e.rekeyFront(key);
+        }
+      }
     }
   }
 }